### PR TITLE
[TT-12266] Add Rate Limit Smoothing Docs (5.4.0)

### PR DESCRIPTION
**Pull request by titpetric**
### For internal users - Please add a Jira DX PR ticket to the subject!
<br>

- Updates rate limiter docs with rate limit smoothing
- Updates key and policy creation examples documenting the value

---

[Preview Link
](https://deploy-preview-4771--tyk-docs.netlify.app/nightly/docs/getting-started/key-concepts/rate-limiting/#rate-limit-smoothing)

<br>

### Description

<br>

#### Screenshots (if appropriate)
<br>

## Checklist

- [x] I have added a **preview link** to the PR description.
- [x] I have [reviewed the guidelines](https://github.com/TykTechnologies/tyk-docs/blob/master/CONTRIBUTING.md) for contributing to this repository.
- [x] I have [read the technical guidelines](https://github.com/TykTechnologies/tyk-docs/blob/master/CONTRIBUTING-TECHNICAL-GUIDE.md) for contributing to this repository.
- [x] Make sure you have started *your change* off *our latest `master`*.
- [x] I **labelled** the PR

___

### **PR Type**
Documentation

___

### **Description**
- Added JSON formatting to API key creation examples.
- Introduced rate limit smoothing configuration in API key creation and security policy creation.
- Updated explanations and examples for new fields in API key creation and security policy creation.
- Added detailed explanation and example configuration for rate limit smoothing in rate limiting concepts.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>create-api-key.md</strong><dd><code>Add rate limit smoothing configuration to API key creation</code></dd></summary>
<hr>

tyk-docs/content/getting-started/create-api-key.md
<li>Added JSON formatting to API key creation examples.<br> <li> Introduced rate limit smoothing configuration in API key creation.<br> <li> Updated explanations for new fields in API key creation.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4771/files#diff-0f2ed66ded035b824ec58588689453de69edef87757c0c0cb87af7b2494f268d">+28/-6</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>create-security-policy.md</strong><dd><code>Add rate limit smoothing configuration to security policy creation</code></dd></summary>
<hr>

tyk-docs/content/getting-started/create-security-policy.md
<li>Added rate limit smoothing configuration to security policy creation.<br> <li> Updated rate limit and quota values in policy examples.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4771/files#diff-916990795f159a580be2bae58f74531b642edb3783d3296d2ab0ef5b2c627809">+17/-9</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rate-limiting.md</strong><dd><code>Add detailed documentation for rate limit smoothing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tyk-docs/content/getting-started/key-concepts/rate-limiting.md
<li>Added detailed explanation of rate limit smoothing.<br> <li> Provided example configuration for rate limit smoothing.<br> <li> Updated Redis Rate Limiter section with new smoothing options.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4771/files#diff-a0f383800801ae765ec875c3c98610f657751ec0779b9e23970357b74fb58e9a">+55/-2</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Review by jeffy-mathew — APPROVED** [review]

File: tyk-docs/content/getting-started/create-api-key.md [hunk 1]
```diff
@@ -1,12 +1,13 @@
 ---
-date: 2017-03-13T15:08:55Z
-Title: Access an API
-tags: ["Tyk Tutorials", "Getting Started", "API Key", "Tyk Cloud", "Tyk Self-Managed", "Tyk Open Source"]
-description: "Creating an API access key using Tyk"
```
Comment (letzya) [inline tyk-docs/content/getting-started/create-api-key.md]: @titpetric any reason for deleting the description?

File: tyk-docs/content/getting-started/create-api-key.md [hunk 2]
```diff
@@ -39,27 +41,28 @@ The value returned in the `key_id` parameter of the response is the access key y
...
 
 **Prerequisite**
 
-*   You will need your API secret, this is the `secret` property of the `tyk.conf` file.
+- You will need your API secret, this is the `secret` property of the `tyk.conf` file.
 
 Once you have this value, you can use them to access the Gateway API, the below `curl` command will generate a key for one of your APIs, remember to replace `{API-SECRET}`, `{API-ID}` and `{API-NAME}` with the real values as well as the `curl` domain name and port to be the correct values for your environment.
 
-```{.copyWrapper}
+```
```
Comment (letzya) [inline tyk-docs/content/getting-started/create-api-key.md]:
```suggestion
```curl
```

File: tyk-docs/content/getting-started/create-api-key.md [hunk 3]
```diff
@@ -74,6 +77,13 @@ curl -X POST -H "x-tyk-authorization: {API-SECRET}" \
     "quota_renews": 1449051461,
     "quota_remaining": -1,
     "quota_renewal_rate": 60,
+    "smoothing": {
```
Comment (letzya) [inline tyk-docs/content/getting-started/create-api-key.md]: @titpetric this comment is important! you only added for open source tab, was that your intention?

Comment (titpetric) [reply tyk-docs/content/getting-started/create-api-key.md]: It was, but we could expand scope if there are more reasonable suggestions? The creating a key and creating a policy pages are central to the new config, but we may also decrease scope and just consider the rate limiting page.

File: tyk-docs/content/getting-started/create-api-key.md (hunk 1 above)
Comment (titpetric) [reply tyk-docs/content/getting-started/create-api-key.md]: Nice catch, this was something between human error and automation error. Re-added description, menu, weight.

**Review by andrei-tyk — APPROVED** [review]

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 4]
```diff
@@ -76,9 +82,60 @@ maintain the request log, mostly impacting CPU usage. Redis resource
...
+The request allowance will be updated internally every `delay` seconds.
+
+Events are emitted based on the following calculations:
+
+- When the request rate rises above `allowance - (step * trigger)`,
+  a RateLimitSmoothingUp event is emitted and allowance increases by `step`.
+- When the request rate falls below `allowance - (step + step * trigger)`,
+  a RateLimitSmoothingDown event is emitted and allowance decreases by `step`.
+
+To see how to configure the value on keys or policies:
```
Comment (dcs3spp) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
Please consult the following guides for an explanation of how to configure the value on keys or policies:
```

File: tyk-docs/content/getting-started/create-security-policy.md [hunk 5]
```diff
@@ -1,20 +1,25 @@
...
+    main:
+        parent: Getting Started
+description: Creating a first security policy for your API using Tyk
+date: "2017-03-13T14:32:26Z"
 aliases:
-  - /getting-started/installation/tutorials/create-security-policy/
-  - /try-out-tyk/tutorials/create-security-policy/
-  - /getting-started/tutorials/create-security-policy/
-  - /try-out-tyk/tutorials/create-security-policy/
+    - /getting-started/installation/tutorials/create-security-policy/
```
Comment (dcs3spp) [inline tyk-docs/content/getting-started/create-security-policy.md]: @titpetric Why the spacing added?

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 6]
```diff
@@ -76,9 +82,60 @@ maintain the request log, mostly impacting CPU usage. Redis resource
...
+```
+
+This is used to compute a request allowance. The request allowance will
+be smoothed between `threshold`, and the defined rate limits (maximum).
+The request allowance will be updated internally every `delay` seconds.
+
+Events are emitted based on the following calculations:
+
+- When the request rate rises above `allowance - (step * trigger)`,
+  a RateLimitSmoothingUp event is emitted and allowance increases by `step`.
```
Comment (dcs3spp) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
  a *RateLimitSmoothingUp* event is emitted and allowance increases by `step`.
```

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 7]
```diff
@@ -76,9 +82,60 @@ maintain the request log, mostly impacting CPU usage. Redis resource
...
+- When the request rate falls below `allowance - (step + step * trigger)`,
+  a RateLimitSmoothingDown event is emitted and allowance decreases by `step`.
```
Comment (dcs3spp) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
  a *RateLimitSmoothingDown* event is emitted and allowance decreases by `step`.
```

**Review by dcs3spp — CHANGES_REQUESTED** [review]
@titpetric PRLGTM I have made some minor suggestions. Why the additional spacing in front matter for aliases?

File: tyk-docs/content/getting-started/create-security-policy.md [hunk 8]
```diff
@@ -51,8 +57,15 @@ Adding a policy to the Tyk Gateway is very easy. Polices are loaded into memory
     },
     "active": true,
     "name": "POLICY NAME",
-    "rate": 100,
+    "rate": 1000,
     "per": 1,
+    "smoothing": {
```
Comment (dcs3spp) [inline tyk-docs/content/getting-started/create-security-policy.md:62]: This does not render correctly in preview, with the `:` highlighted

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 9]
```diff
@@ -76,9 +82,60 @@ maintain the request log, mostly impacting CPU usage. Redis resource
...
+
+Rate Limit Smoothing is configured using:
+
+- API Keys, under `.smoothing.*`
+- Policies, under `.smoothing.*`
+- Per-API, via Keys and Policies under `.access_rights[<api_id>].limit.smoothing.*`
+
+An example configuration would be as follows:
+
+```json
```
Comment (dcs3spp) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
```yaml
```

**Comment by dcs3spp** [comment]
@titpetric Is there rate limiting content that needs to be added to Self Managed and Cloud tabs for [create security policies](https://deploy-preview-4771--tyk-docs.netlify.app/docs/nightly/getting-started/create-security-policy/) and [access api keys](http://localhost:1313/docs/nightly/getting-started/create-api-key/)?

File: tyk-docs/content/getting-started/create-api-key.md [hunk 10]
```diff
@@ -1,20 +1,26 @@
...
+date: "2017-03-13T15:08:55Z"
 aliases:
-  - /with-tyk-community-edition/tutorials/create-api-token/
-  - /get-started/with-tyk-multi-cloud/tutorials/create-api-token/
-  - /get-started/with-tyk-on-premise/tutorials/tyk-on-premise-pro/create-api-token/
-  - /get-started/with-tyk-cloud/tutorials/create-api-token/
-  - /try-out-tyk/tutorials/create-api-key/
-  - /try-out-tyk/create-api-key/
-  - /getting-started/tutorials/create-api-key/
+    - /with-tyk-community-edition/tutorials/create-api-token/
```
Comment (letzya) [inline tyk-docs/content/getting-started/create-api-key.md]: @titpetric your ide keeps adding these. Is 4 spaces a tyk standard in the code?

File: tyk-docs/content/getting-started/create-api-key.md [hunk 11]
```diff
@@ -23,7 +25,8 @@ aliases:
 {{< include "create-api-key-include" >}}
 
 You will see a 200 response with your new key:
-```
+
+```json
```
Comment (letzya) [inline tyk-docs/content/getting-started/create-api-key.md]:
```suggestion
```yaml
```

File: tyk-docs/content/getting-started/create-api-key.md [hunk 12]
```diff
@@ -85,9 +96,22 @@ curl -X POST -H "x-tyk-authorization: {API-SECRET}" \
...
+The above creates a new key with the rate limits, and security profile that grants access to the APIs listed in the `access_rights` section.
+
+- `{API-ID}`: The API ID you wish this policy to grant access to, there can be more than one of these entries.
+- `{API-NAME}`: The name of the API being granted access to (this is not required, but helps when debugging or auditing).
+
+The important elements:
+
+- `access_rights`: A list of objects representing which APIs you have configured to grant access to.
+- `rate` and `per`: The number of allowed requests per period.
+- `smoothing`: The Rate Limit Smooting configuration for Redis Rate Limiter.
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/create-api-key.md]:
```suggestion
```

File: tyk-docs/content/getting-started/create-security-policy.md [hunk 13]
```diff
@@ -62,17 +75,17 @@ Adding a policy to the Tyk Gateway is very easy. Polices are loaded into memory
...
 
 The important elements:
 
-*   `access_rights`: A list of objects representing which APIs that you have configured to grant access to.
-*   `rate` and `per`: The number of requests to allow per period.
-*   `quota_max`: The maximum number of allowed requests over a quota period.
-*   `quota_renewal_rate`: how often the quota resets, in seconds. In this case we have set it to renew every hour.
+- `access_rights`: A list of objects representing which APIs that you have configured to grant access to.
+- `rate` and `per`: The number of requests to allow per period.
+- `smoothing`: The Rate Limit Smooting configuration for Redis Rate Limiter.
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/create-security-policy.md:86]:
```suggestion
- `smoothing`: The Rate Limit Smoothing configuration for Redis Rate Limiter.
```

**Review by andyo-tyk — CHANGES_REQUESTED** [review]
Corrected typos

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 14]
```diff
@@ -1,41 +1,104 @@
...
 
-1. Distributed Rate Limiter. Recommended for most use cases. Implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket).
-2. Redis Rate Limiter. Implements the [sliding window log algorithm](https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window).
-3. Fixed Window Rate Limiter. Implements the [fixed window algorithm](https://redis.io/learn/develop/dotnet/aspnetcore/rate-limiting/fixed-window).
+Tyk offers the following rate limiting algorithms:
 
-When the rate limits are reached, Tyk will block request with a [429 (Rate Limit Exceeded)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) response.
+1. [Distributed Rate Limiter]({{< ref "#Distributed-Rate-Limiter" >}}): recommended for most use cases, implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket)
+2. [Redis Rate Limiter]({{< ref "#Redis-Rate-Limiter" >}}): implements the [sliding window log algorithm](https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window)
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
2. [Redis Rate Limiter]({{< ref "#redis-rate-limiter" >}}): implements the [sliding window log algorithm](https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window)
```

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 15]
```diff
@@ -1,41 +1,104 @@
...
 
-1. Distributed Rate Limiter. Recommended for most use cases. Implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket).
-2. Redis Rate Limiter. Implements the [sliding window log algorithm](https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window).
-3. Fixed Window Rate Limiter. Implements the [fixed window algorithm](https://redis.io/learn/develop/dotnet/aspnetcore/rate-limiting/fixed-window).
+Tyk offers the following rate limiting algorithms:
 
-When the rate limits are reached, Tyk will block request with a [429 (Rate Limit Exceeded)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) response.
+1. [Distributed Rate Limiter]({{< ref "#Distributed-Rate-Limiter" >}}): recommended for most use cases, implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket)
```
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]:
```suggestion
1. [Distributed Rate Limiter]({{< ref "#distributed-rate-limiter" >}}): recommended for most use cases, implements the [token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket)
```

**Review by andyo-tyk — CHANGES_REQUESTED** [review]
Correcting local links

File: tyk-docs/content/getting-started/create-api-key.md (hunk 3 above)
Comment (andyo-tyk) [reply tyk-docs/content/getting-started/create-api-key.md]: I would suggest for now we don't include the smoothing controls in these getting started guides, as it's an advanced feature.

Comment (letzya) [reply tyk-docs/content/getting-started/create-api-key.md]: good point @andyo-tyk. @dcs3spp we can just remove it and merge

File: tyk-docs/content/getting-started/create-security-policy.md (hunk 8 above)
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/create-security-policy.md]: I wouldn't include this

File: tyk-docs/content/getting-started/create-security-policy.md (hunk 13 above)
Comment (andyo-tyk) [inline tyk-docs/content/getting-started/create-security-policy.md]:
```suggestion
```

**Review by andyo-tyk — DISMISSED** [review]
Recommend not making changes to the getting started guides at this stage.

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md (hunk 6 above)
Comment (andyo-tyk) [reply tyk-docs/content/getting-started/key-concepts/rate-limiting.md]: This is no longer relevant following re-write of the page. Closing comment.

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md (hunk 7 above)
Comment (andyo-tyk) [reply tyk-docs/content/getting-started/key-concepts/rate-limiting.md]: This is no longer relevant following re-write of the page. Closing comment.

**Review by dcs3spp — APPROVED** [review]

**Comment by dcs3spp** [comment]
@titpetric I have approved the PR and removed the rate limit smoothing in the examples. @andyo-tyk are you ok with the changes? If so I will merge ready for release

File: tyk-docs/content/getting-started/create-api-key.md (hunk 1 above)
Comment (titpetric) [reply tyk-docs/content/getting-started/create-api-key.md]: @dcs3spp Is the menu now removed on purpose? I remember re-adding it in, but now doesn't exist in current state of the page

File: tyk-docs/content/getting-started/create-security-policy.md [hunk 16]
```diff
@@ -1,20 +1,25 @@
...
+  - Getting Started
+  - Security Policy
+  - Tyk Cloud
+  - Tyk Self-Managed
+  - Tyk Open Source
 menu:
-    main: 
-      parent: "Getting Started"
-weight: 5
+    main:
```
Comment (titpetric) [inline tyk-docs/content/getting-started/create-security-policy.md]: Still missed indentation on `menu` (even after correcting the one in tags)

File: tyk-docs/content/getting-started/key-concepts/rate-limiting.md [hunk 17]
```diff
@@ -1,41 +1,104 @@
...
-date: 2021-02-04
-tags: ["Rate Limit", "Rate Limiting", "Rate Limit Algorithms", "DRL"]
-description: "Overview of Rate Limiting with the Tyk Gateway"
-menu:
-  main:
-    parent: "Key Concepts"
-weight: 65
+title: Rate Limiting in Tyk
+tags:
+    - Rate Limit
```
Comment (titpetric) [inline tyk-docs/content/getting-started/key-concepts/rate-limiting.md]: Whitespace

File: tyk-docs/content/getting-started/create-api-key.md (hunk 1 above)
Comment (titpetric) [reply tyk-docs/content/getting-started/create-api-key.md]: I've reverted all front matter

File: tyk-docs/content/getting-started/create-security-policy.md (hunk 5 above)
Comment (titpetric) [reply tyk-docs/content/getting-started/create-security-policy.md]: reverted, was added from task md:fmt (https://github.com/bwplotka/mdox/issues/154)

**Comment by titpetric** [comment]
@dcs3spp I've asked @jeffy-mathew to re-approve as a significant section for rate limiting docs was rewritten, which was out of scope for the ticket.

File: tyk-docs/content/getting-started/create-security-policy.md (hunk 16 above)
Comment (dcs3spp) [reply tyk-docs/content/getting-started/create-security-policy.md]: Thanks @titpetric fixed

**Comment by dcs3spp** [comment]
> @dcs3spp I've asked @jeffy-mathew to re-approve as a significant section for rate limiting docs was rewritten, which was out of scope for the ticket.

@titpetric Jeffy has approved, can this now be merged ready for release?

**Comment by titpetric** [comment]
@dcs3spp that's a stale approval, he needs to review the changes and re-approve due to the rewrites.

**Comment by dcs3spp** [comment]
@jeffy-mathew I have re-requested a peer review on behalf of @titpetric

**Review by jeffy-mathew — APPROVED** [review]

**Comment by jeffy-mathew** [comment]
@titpetric @dcs3spp lgtm, I've approved.

**Review by andyo-tyk — APPROVED** [review]

**Comment by dcs3spp** [comment]
Many thanks @jeffy-mathew I will merge ready for release

**Pull request merged**